File: packages/llm-test-harness/src/metrics/contains.ts
import type { MetricFn, MetricScore } from '../types.js'

export class Contains implements MetricFn {
  readonly name = 'Contains'

  constructor(
    private substring: string,
    private caseSensitive = true,
  ) {}

  evaluate(text: string): MetricScore {
    const haystack = this.caseSensitive ? text : text.toLowerCase()
    const needle = this.caseSensitive ? this.substring : this.substring.toLowerCase()
    const pass = haystack.includes(needle)
    return { name: this.name, pass, score: pass ? 1 : 0 }
  }
}

export class ContainsAll implements MetricFn {
  readonly name = 'ContainsAll'

  constructor(
    private substrings: string[],
    private caseSensitive = true,
  ) {}

  evaluate(text: string): MetricScore {
    const haystack = this.caseSensitive ? text : text.toLowerCase()
    const found = this.substrings.filter((s) => {
      const needle = this.caseSensitive ? s : s.toLowerCase()
      return haystack.includes(needle)
    })
    const score = this.substrings.length === 0 ? 1 : found.length / this.substrings.length
    const pass = score === 1
    const missing = this.substrings.filter((s) => {
      const needle = this.caseSensitive ? s : s.toLowerCase()
      return !haystack.includes(needle)
    })
    return {
      name: this.name,
      pass,
      score,
      reason: missing.length > 0 ? `Missing: ${missing.map((s) => `"${s}"`).join(', ')}` : undefined,
    }
  }
}

File: packages/llm-test-harness-python/llm_test_harness/metrics/contains.py
from __future__ import annotations

from ..types import MetricFn, MetricScore


class Contains(MetricFn):
    def __init__(self, substring: str) -> None:
        self._substring = substring

    @property
    def name(self) -> str:
        return "Contains"

    def evaluate(self, text: str) -> MetricScore:
        passed = self._substring in text
        return MetricScore(
            name=self.name,
            passed=passed,
            score=1.0 if passed else 0.0,
        )


class ContainsAll(MetricFn):
    def __init__(self, substrings: list[str]) -> None:
        self._substrings = substrings

    @property
    def name(self) -> str:
        return "ContainsAll"

    def evaluate(self, text: str) -> MetricScore:
        if not self._substrings:
            return MetricScore(name=self.name, passed=True, score=1.0)
        found = sum(1 for s in self._substrings if s in text)
        score = found / len(self._substrings)
        passed = found == len(self._substrings)
        reason = None if passed else f"{found}/{len(self._substrings)} substrings found"
        return MetricScore(name=self.name, passed=passed, score=score, reason=reason)
